### PR TITLE
Error handling - Correlation ID

### PR DESCRIPTION
**Pull request by yedidyas**
I couldn't find any documentation for correlation ID practice, so here it is. LMK what you think.

@goldbergyoni

### CODE REVIEW

File: sections/errorhandling/correlationid.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Correlation ID: help your logs tell you a story and give your error a context
+
+### One Paragraph Explainer
+
+Correlation ID is one of the best problem-solving patterns. It lets you link log records, even if they belong to different services. If your system consumes other services and is itself a producer service, adding a correlaction ID is a must. By this pattern, your transaction logs can become into a story that tells itself by filtering your logs with a specific correlation ID, instead of try linking the cross-services transaction logs to each other by yourself. It can save your day when a process including 20 different microservices throws an exception in one of them, and you have no idea where did the problem start across the flow.
```
Comment (goldbergyoni) [inline sections/errorhandling/correlationid.md:5]: Great writing.

I suggest that the 'microservice' thing should not be the primary point rather secondary (less people do Microservice than we thing), where the main point is correlating entries from single Node.js process across async functions. For example, how to tell this sentence from the api/controller layer belongs to the same request as that entry from the DAL

File: sections/errorhandling/correlationid.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+app.use((req, res, next) => {
+  // Extract the correlation ID from the previous request, or generate it if this is the first request in the transaction
+  const correlationId = req.get('X-Correlation-ID') || uuid.v4();
+  
+  // Set the correaltion ID on the http context
+  httpContext.set('correlationId', correlationId);
+  
+  // Set the correaltion ID on the response
+  res.set('X-Correlation-ID', correlationId);
```
Comment (goldbergyoni) [inline sections/errorhandling/correlationid.md:24]: Clean code.

The main point is left out - Show logger code that reads and prints this .

Another example - Show http-request that includes this.

File: sections/errorhandling/correlationid.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  next();
+});
+```
+<br/>
+
+### Blog Quote: "The notion of a Correlation ID is simple. It’s a value that is common to all requests, messages and responses in a given transaction. With this simplicity you get a lot of power."
+
+From [rapid7](https://blog.rapid7.com/2016/12/23/the-value-of-correlation-ids/)
+
+> When you can group a transaction’s events under a unifying value, the Correlation ID, you can spend your time fixing the problem rather than finding the problem. Without a Correlation ID, well… there is a saying, “if you put a chimpanzee in front of typewriter for eternity, eventually the animal will type out Hamlet.” The same could be said for logging without a Correlation ID, except none of us will be here for eternity. Use a Correlation ID. You will be glad you did.
```
Comment (goldbergyoni) [inline sections/errorhandling/correlationid.md:35]: > “if you put a chimpanzee in front of typewriter for eternity, eventually the animal will type out Hamlet.” The same could be said for logging without a Correlation ID, except none of us will be here for eternity. Use a Correlation ID. You will be glad you did.

Funny thing, but I didn't fully get the correlation between repetitive writing and correlation-id?

File: sections/errorhandling/correlationid.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+Correlation ID is one of the best problem-solving patterns. It lets you link log records, even if they belong to different services. If your system consumes other services and is itself a producer service, adding a correlaction ID is a must. By this pattern, your transaction logs can become into a story that tells itself by filtering your logs with a specific correlation ID, instead of try linking the cross-services transaction logs to each other by yourself. It can save your day when a process including 20 different microservices throws an exception in one of them, and you have no idea where did the problem start across the flow.
+
+<br/>
+
+### Code Example: passing the correlation ID between services on the requet http context
+Here is an example of using [express-http-context](https://www.npmjs.com/package/express-http-context) library as part of a middleware to set the forwarded correlation ID on the http context after it was extracted from the request. 
+If the correlation ID doesn't exist on the request, since this may be the first request in the requests chain, it's being generated.
+
+```javascript
+const httpContext = require('express-http-context');
```
Comment (goldbergyoni) [inline sections/errorhandling/correlationid.md:14]: Node now has a built-in mechanism, async-local-storage which is also not express specific (we try to move away from being too express-centric). Shall we use this?

File: sections/errorhandling/correlationid.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  
+  // Set the correaltion ID on the response
+  res.set('X-Correlation-ID', correlationId);
+
+  next();
+});
+```
+<br/>
+
+### Blog Quote: "The notion of a Correlation ID is simple. It’s a value that is common to all requests, messages and responses in a given transaction. With this simplicity you get a lot of power."
```
Comment (goldbergyoni) [inline sections/errorhandling/correlationid.md:31]: Just an idea - I would put here an image of some log aggregator UI (e.g. ELK) that shows how the 'story' looks like. I can get one

**Review by goldbergyoni — CHANGES_REQUESTED** [review]
@yedidyas Great writing, a promising start ✨.

Note that adding a new BP is a multi back and forth journey where we refine it over a long discussion until it really stands out. For me, maybe also for you, it's an opportunity to know this strategic topic much better.

I left multiple comments, there is one fundamental issue: We already have a similar [BP 5.14 ](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/production/assigntransactionid.md) which is filed under 'Going to production'. I also believe that it's the right place since we want to correlate not only error rather the entire story, observability is not about errors, the worst challenges happen when there is no exception... Shall we do our work over that item?

File: sections/errorhandling/correlationid.md (hunk 3 above)
Comment (yedidyas) [reply sections/errorhandling/correlationid.md:35]: As I understand it, he means that logs without correlation id are not readable exactly like the output of a typing chimpanzee is not readable ... not so smart as it tries to be heard :)

File: sections/errorhandling/correlationid.md (hunk 5 above)
Comment (yedidyas) [reply sections/errorhandling/correlationid.md:31]: I've just found in our logs this example- 
https://drive.google.com/file/d/1AYdmHhSZGcJaGIFA1pdMX15NMwG-R3de/view?usp=sharing
(microservices real names and data are edited)
Is it the kind of story to show in an image you thought about?

File: sections/errorhandling/correlationid.md (hunk 4 above)
Comment (yedidyas) [reply sections/errorhandling/correlationid.md:15]: Very interesting - I didn't know about this ability.
I'm not sure it should completely replace the current example; async-local-storage is useful only for the async functions use case you mentioned.
The multiple-microservices use case can't be solved by that async-local-storage solution...

**Comment by yedidyas** [comment]
> @yedidyas Great writing, a promising start ✨.
> 
> Note that adding a new BP is a multi back and forth journey where we refine it over a long discussion until it really stands out. For me, maybe also for you, it's an opportunity to know this strategic topic much better.
> 
> I left multiple comments, there is one fundamental issue: We already have a similar [BP 5.14 ](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/production/assigntransactionid.md) which is filed under 'Going to production'. I also believe that it's the right place since we want to correlate not only error rather the entire story, observability is not about errors, the worst challenges happen when there is no exception... Shall we do our work over that item?
> 
> I know that complicates stuff, but it's an opportunity to dive deeper into this topic, and for us to really master it

I really didn't understand how can it be that there's no mention for correlation id in the repo, now it makes sense, it already existed :) missed that. Isn't correlation id the "formal" name for this pattern?

It's a really interesting topic.
I will keep reading about it - and will open a new PR suggesting some improvements to BP 5.14, regarding your comments here.
Thanks!

File: sections/errorhandling/correlationid.md (hunk 3 above)
Comment (kevynb) [reply sections/errorhandling/correlationid.md:36]: I agree with @goldbergyoni, I have troubles finding a *correlation* (😉 ) between the two. I think we could dispense from this quote :)

File: sections/errorhandling/correlationid.md (hunk 4 above)
Comment (kevynb) [reply sections/errorhandling/correlationid.md:15]: Using 'cls-hooked' and 'cls-rtracer' works very well for the use case you are describing. With multiple microservices one just has to add an http header with the correlationId on outgoing calls so that it's picked by the next MS.

Those plugin uses the `async-local-storage` api under the hood and provide middleware for well-known frameworks like express, fastify, hapi, koa

File: sections/errorhandling/correlationid.md (hunk 5 above)
Comment (kevynb) [reply sections/errorhandling/correlationid.md:32]: I think showing images could be great to help people understand the concept better!

Do you think you could also generate an image with unrelated logs? To show the before and after?

I think some cropping around the log area would help the reader to focus

**Comment by yedidyas** [comment]
@goldbergyoni  @kevynb 
I've just applied all the changes you suggested in the comments.
When will be relevant, after merging to master, the images URL will be replaced with the local repo URLs.

Not sure if I dived too much into the async-local-storage issue, but I've decided to add its restrictions and due to that, leave the old example in addition to the new one.

Have a nice weekend!

**Review by elektronik2k5 — DISMISSED** [review]

File: sections/errorhandling/correlationid.md (hunk 4 above)
Comment (goldbergyoni) [reply sections/errorhandling/correlationid.md]: @kevynb You sure that cls-hooked is using async-local-storage? Why use this and not async-local-storage directly?

File: sections/errorhandling/correlationid.md (hunk 5 above)
Comment (goldbergyoni) [reply sections/errorhandling/correlationid.md]: Yes, I also like @kevynb idea of showing an 'Anti-pattern' image that shows a messy log

File: sections/errorhandling/correlationid.md (hunk 4 above)
Comment (yedidyas) [reply sections/errorhandling/correlationid.md]: @goldbergyoni 
Check my last commits - cls-rtracer totally simplifies the process.
I changed the whole PR

**Review by goldbergyoni — CHANGES_REQUESTED** [review]
Great improvements.

One fundamental thought - The main example should not rely on a niche lib that has 150 GH stars. Start with asyncLocalStorage example, let the reader understand the low-level API which not too complicated (i.e. 6 loc), then the 2nd example might be 'Example: Using helper library to simplify the syntax', use rtracer there

Other minor suggestions - 

1. Improve the home page title - You didn't find this item, other also probably missed it, this content is associated with multiple terms like 'correlation id', 'request id', 'request context'. Consider adding to the home page something like 'also known as correlation-id, request context, etc'

2. Bonus: Could be awesome to include a Fastify or Nest example
https://github.com/fastify/fastify-request-context

They will appreciate it that we start showing example for their platform and help promoting/tweeting this enhanced content

File: sections/errorhandling/correlationid.md (hunk 4 above)
Comment (kevynb) [reply sections/errorhandling/correlationid.md]: @goldbergyoni Yes, they say it on the repo homepage and you can find it if you dig into the code

File: sections/production/assigntransactionid.md [hunk 6]
```diff
@@ -8,7 +8,65 @@ A typical log is a warehouse of entries from all components and requests. Upon d
...
+### Code example: sharing TransactionId among current request functions 
+using [cls-rtracer](https://www.npmjs.com/package/cls-rtracer) (a library based on [async-local-storage](https://nodejs.org/api/async_hooks.html#async_hooks_class_asynclocalstorage), implemented for Express & Koa middlewares and Fastify & Hapi plugins)
+
+```javascript
+const express = require('express');
+const rTracer = require('cls-rtracer');
+
+const app = express();
+
+app.use(rTracer.expressMiddleware());
```
[the remark this answers is not in the document]
Comment (kevynb) [reply sections/production/assigntransactionid.md:20]: Yes, it's been running on a production system for 6 months without issue. Anyway I agree with your statements. It's better if people understand what they are doing, too much magic is not good when you have a problem. 
Let's do as you suggest

File: sections/production/assigntransactionid.md [hunk 7]
```diff
@@ -8,7 +8,65 @@ A typical log is a warehouse of entries from all components and requests. Upon d
...
+    // Respect TransactionId from header
+    useHeader: true,
+    // TransactionId header name
+    headerName: 'x-transaction-id'
+}));
+
+const axios = require("axios");
+
+// Now, the external service will automaticlly get the current TransactionId as header
+const response = await axios.get('https://externalService.com/api/getAllUsers');
```
[the remark this answers is not in the document]
Comment (kevynb) [reply sections/production/assigntransactionid.md:57]: I agree, even though this library is really handy, it's better to learn by doing it by yourself at first

File: sections/production/assigntransactionid.md [hunk 8]
```diff
@@ -8,7 +8,65 @@ A typical log is a warehouse of entries from all components and requests. Upon d
...
+```
+<br/><br/>
+
+
+There are two restrictions on using cls-rtracer, due to its dependence on async-local-storage (You can think of async-local-storage as the node.js alternative to thread local storage, more details [here](https://www.freecodecamp.org/news/async-local-storage-nodejs/)):
+1. async-local-storage requires Node v.14. 
+2. async-local-storage is based on a lower level construct in Node called async_hooks which is still experimental, so you may have the fear of performance problems. Even if they do exist, they are very negligible, but you should make your own considerations.
```
Comment (kevynb) [inline sections/production/assigntransactionid.md:64]: I think those points should be kept as they apply to the underlying `async-local-storage` even if we remove `cls-rtracer` form the main example.

File: sections/production/assigntransactionid.md [hunk 9]
```diff
@@ -37,3 +95,21 @@ class logger {
...
+</details>
+
+<br/><br/>
+
+### Good: Logs with an assigned TransactionId - can be used as filter to see only a single flow
+![alt text](https://i.ibb.co/YjJwgbN/logs-with-transaction-id.jpg "Logs with transaction id")
+<br/><br/>
+
+### Bad: logs without a TransactionId - no option to use a filter and see only a single flow, you need to understand by yourself which logs are relevant between all the "noise" around
+![alt text](https://i.ibb.co/PFgVNfn/logs-withtout-transaction-id.jpg "Logs with transaction id")
```
[the remark this answers is not in the document]
Comment (kevynb) [reply sections/production/assigntransactionid.md:107]: I like how you did the blur to show that the one you want are lost in the middle of the others! 👀

**Review by kevynb — COMMENTED** [review]
Minor comments. I like the pics you added a lot! Sorry for sending you on a path with `cls-tracer` and then changing my mind after yoni's remarks 😖

**Comment by goldbergyoni** [comment]
@yedidyas Just ensuring that you are not waiting for some input from my end. Take your time of course, this is OSS land - a planet without clock and gravity:)

We are one step shy from a great improvement - Just put a asyncLocalStorage first, then the great content that you added

**Comment by yedidyas** [comment]
@goldbergyoni 
Sure, I'm on it. had some busy days. I will let you know once you can review again. thanks :)

File: README.md [hunk 10]
```diff
@@ -698,7 +698,9 @@ All statements above will return false if used with `===`
 
 ## ![✔] 5.14. Assign a transaction id to each log statement
 
-**TL;DR:** Assign the same identifier, transaction-id: {some value}, to each log entry within a single request. Then when inspecting errors in logs, easily conclude what happened before and after. Unfortunately, this is not easy to achieve in Node due to its async nature, see code examples inside
+Also known as correlation id / transit id / tracing id / request id / request context / etc.
+
+**TL;DR:** Assign the same identifier, transaction-id: {some value}, to each log entry within a single request. Then when inspecting errors in logs, easily conclude what happened before and after. Until version 14 of Node, this was not easy to achieve due to Node's async nature, but since AsyncLocalStorage came to town, this became possible and easy than ever. see code examples inside
```
Comment (elektronik2k5) [inline README.md:703]:
```suggestion
**TL;DR:** Assign the same identifier, transaction-id: {some value}, to each log entry within a single request. Then when inspecting errors in logs, easily conclude what happened before and after. Until version 14 of Node, this was not easy to achieve due to Node's async nature, but since `AsyncLocalStorage` came to town, this became possible and easy than ever. See code examples inside.
```

File: sections/production/assigntransactionid.md [hunk 11]
```diff
@@ -6,10 +6,80 @@
...
+        console.log(`${message} ${asyncLocalStorage.getStore().get('transactionId')}`);
+    }
+}
+```
+<br/>
+
+<details>
+<summary><strong>Code example: Using helper library to simplify the syntax</strong></summary>
+
+Sharing TransactionId among current request functions using [cls-rtracer](https://www.npmjs.com/package/cls-rtracer) (a library based on async-local-storage, implemented for Express & Koa middlewares and Fastify & Hapi plugins)
```
Comment (elektronik2k5) [inline sections/production/assigntransactionid.md:82]:
```suggestion
Sharing TransactionId among current request functions using [cls-rtracer](https://www.npmjs.com/package/cls-rtracer) - a library based on async-local-storage, implemented for Express & Koa middlewares and Fastify & Hapi plugins.

```

**Comment by yedidyas** [comment]
@goldbergyoni  I pushed all my changes

**Comment by goldbergyoni** [comment]
@yedidyas Thanks a bunch. I blocked the entire Friday morning for OSS work and this is my first task as I'm having the morning coffee ☕️😉

**Review by goldbergyoni — APPROVED** [review]
LGTM

**Comment by goldbergyoni** [comment]
@yedidyas Great content and thanks for being patient during these not-short improvements iterations. I think that the content now with the bare-naked example and pictures is really exhaustive! Should someone wishes to learn about this topic in Node.js, this is now probably one of the best options, if not the best.

I'll tweet about this great improvement and put it in the news section next week (have Twitter handle?).

Let's stay in touch.

 @all-contributors please add @yedidyas  for content

**Comment by yedidyas** [comment]
@goldbergyoni Awesome :) I'm happy to be a part of this amazing project.
That's me on tweeter:
https://twitter.com/yedidya_s
